common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/QueryContentProviderLoader.kt

```diff
@@ -0,0 +1,102 @@
+package com.microsoft.identity.common.internal.broker.ipc
+
+import android.content.Context
+import android.content.pm.PackageManager
+import android.content.pm.ProviderInfo
+import android.os.Build
+import android.os.DeadObjectException
+import com.microsoft.identity.common.java.interfaces.INameValueStorage
+import com.microsoft.identity.common.java.interfaces.IPlatformComponents
+import com.microsoft.identity.common.logging.Logger
+
+/**
+ * packageManager.queryContentProviders involves an ipc call (we've seen DeadObjectException from it before)
+ *
+ * We'll avoid making unnecessary ipc request by caching them.
+ * The content provider support status should remain the same for a given broker version.
+ **/
+class QueryContentProviderLoader(
+    private val getVersionCode: (appPackageName: String) -> String,
+    private val supportedByContentProvider: (appPackageName: String) -> Boolean,
+    private val fileManager: INameValueStorage<String>,
+) : IQueryContentProviderLoader {
+
+    companion object {
+        private val TAG = QueryContentProviderLoader::class.java.simpleName
+        private const val SHARED_PREFERENCE_NAME = "com.microsoft.common.ipc.content.provider.query.cache"
+
+        @Throws(PackageManager.NameNotFoundException::class)
+        private fun getVersionCode(context: Context, appPackageName: String) : String {
+            val packageInfo = context.packageManager.getPackageInfo(appPackageName, 0)
+            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                packageInfo.longVersionCode.toString()
+            } else {
+                packageInfo.versionCode.toString()
+            }
+        }
+
+        @Throws(DeadObjectException::class)
+        fun supportedByContentProvider(context: Context, appPackageName: String) : Boolean {
+            val contentProviderAuthority: String =
+                ContentProviderStrategy.getContentProviderAuthority(appPackageName)
+
+            val providers: List<ProviderInfo> = context.packageManager
+                .queryContentProviders(null, 0, 0)
+
+            for (providerInfo in providers) {
+                if (providerInfo.authority != null && providerInfo.authority == contentProviderAuthority) {
+                    return true
+                }
+            }
+
+            return false
+        }
+    }
+
+    constructor(context: Context, components: IPlatformComponents): this(
+        fileManager = components.storageSupplier.getUnencryptedNameValueStore(
+            SHARED_PREFERENCE_NAME, String::class.java),
+        getVersionCode = { pkgName -> getVersionCode(context, pkgName) },
+        supportedByContentProvider = { pkgName -> supportedByContentProvider(context, pkgName) }
+    )
+
+    override fun getResult(packageName: String) : Boolean {
+        val methodTag = "$TAG:getResult"
+        try {
+            val key = packageName + ":" + getVersionCode(packageName)
+
+            fileManager.get(key)?.let {
```

Comment:
Is this doing an actual file read?
there would be some io cost associated with file reads, which may add to the overall perf.

Reply:
File read. 

It's IO either way though, given that the actual operation seems to involve some sort of ipc call.

The pros of this is that we would decrease the load on the shared ipc buffer. (1mb per process, which could add up)

Reply:
This IPC buffer is shared across all apps on the device?

Reply:
1MB per process. (not much official docs on this here. This is based on the [ContentProvider design docs we have)](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/BrokerContract/ContentProviderAndroidBroker.md&_a=preview)